### PR TITLE
add some considerations for collections

### PR DESCRIPTION
**Pull request by richm**
add some considerations for collections for modules and "private" roles

### CODE REVIEW

**Comment by richm** [comment]
@greg-hellings this came up as part of the conversion of system roles to collections format
@nhosoi @pcahyna ptal

File: README.md [hunk 1]
```diff
@@ -513,6 +513,32 @@ and [development](https://docs.ansible.com/ansible/latest/dev_guide/index.html).
...
 
+## Collections
+
+### Modules and module_utils
+In a collection, the names of the module and module_utils files must be unique
+since all of them are located in
+`COLLECTION_NAMESPACE/COLLECTION_NAME/plugins/modules` (modules) and
+`COLLECTION_NAMESPACE/COLLECTION_NAME/plugins/module_utils` (module supporting
+files).  It is recommended to:
+* start module filenames with `ROLENAME_`
```
Comment (greg-hellings) [inline README.md:524]: I'm not sure I agree with this. One of the benefits of a collection is that the modules, plugins, filters, and so forth no longer need to be bound up to a role. The name should be indicative of WHAT it does, not WHERE it does it. It easily could be the case that a module or filter is useful to far more than just the role it was originally developed for.

File: README.md [hunk 2]
```diff
@@ -513,6 +513,32 @@ and [development](https://docs.ansible.com/ansible/latest/dev_guide/index.html).
...
+* use a sub-directory `module_utils/ROLENAME/` for the role-specific
+  module_utils files
+* avoid generic names such as `size.py`
+
+If you are converting several roles into collection format and bundling them in
+a collection, you may need to rename files so that they are unique.  You may
+have to edit your module files to adjust the python imports.
```
Comment (greg-hellings) [inline README.md:531]: True, you may have to rename them. But, if you have several roles that belong in a single collection because they are related and multiple of them have the same or similarly named python imports, then perhaps you're the perfect example for why such an import should NOT be prefaced with the role name where it is used.

File: README.md [hunk 3]
```diff
@@ -513,6 +513,32 @@ and [development](https://docs.ansible.com/ansible/latest/dev_guide/index.html).
...
+
+### Private Roles
+If your collection uses "private" roles - roles used to implement features in
+the public roles, but not meant to be used outside of the role - these roles are
+"top level" roles in the collection, meaning they can be used directly by users - there is no mechanism in Ansible to restrict role usage.  If you want to denote
```
Comment (greg-hellings) [inline README.md:536]: This line is too long.

File: README.md [hunk 4]
```diff
@@ -513,6 +513,32 @@ and [development](https://docs.ansible.com/ansible/latest/dev_guide/index.html).
...
+these roles as being "private" to the collection, and not supported for use
+directly, begin the name of the role with an underscore `_`.  This doesn't make
```
Comment (greg-hellings) [inline README.md:538]: This is unpossible. Ansible roles in a collection MUST adhere to the naming convention that they consist only of alphanumerics and the underscore character and they MUST begin with an alpha character. Galaxy will fail the import of the collection if this suggestion is implemented. The rules against this are hard coded into the Ansible Galaxy code checker and into ansible-lint to ensure that exactly this does not happen.

**Review by greg-hellings — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (greg-hellings) [reply README.md:524]: See also: a collection can exist solely for the benefit of distributing modules. "theforeman.foreman" is just such a collection. It consists entirely of module and moule_util imports and not a single role to be seen.

Comment (richm) [reply README.md:524]: Ok.  I'll reword this to be specifically about the case where you are converting roles with modules to collections format.

File: README.md (hunk 2 above)
Comment (richm) [reply README.md:531]: Not sure I get your point

File: README.md (hunk 4 above)
Comment (richm) [reply README.md:538]: Hmm - because it actually works to have roles in the collection with a name that begins with an underscore.  At least, we have been doing integration testing - we have not run ansible-lint against the converted collection, nor have we tried to import into Galaxy.  So it would seem the checkers are more restrictive than Ansible itself in this respect.
Do you have a link to the documentation where it says that the names of the roles in a collection must begin with alphabetic?  @nhosoi @pcahyna looks like we'll have to scrap that feature of the system roles conversion, and just document that the rsyslog and pcp roles should not be used directly.

Comment (greg-hellings) [reply README.md:538]: Here is the relevant bit from ansible-lint: https://github.com/ansible/ansible-lint/blob/master/lib/ansiblelint/rules/RoleNames.py#L29

Ansible-lint is part of the Galaxy import. So you're right - there's nothing in the Ansible engine that prevents underscores in the role names but you should find an error when you run Ansible Lint and Galaxy will fail on the import process.

File: README.md (hunk 1 above)
Comment (greg-hellings) [reply README.md:524]: Even then I would not encourage naming modules after the roles they came from. Leave their names as something that indicates what they do and leave it at that.

File: README.md (hunk 2 above)
Comment (greg-hellings) [reply README.md:531]: You're saying that it's possible you import roles A, B, C and all have a module named `foo`. You're going to need to rename at least some of them when they join the collection. That's true.

However, this is a great argument against your point above asking for them all to be named A_foo, B_foo, C_foo. If they're all named `foo` that's probably because there's some shared functionality or similarity at least that needs to be shared more broadly than just within a particular role. So either they need to be named differently (e.g. "size" is very ambiguous) or they should NOT be named after their original role because this is clearly functionality that is needed for more than just that one specific role. So just name it `foo` and make sure it covers all the use cases in the roles.

**Comment by richm** [comment]
hmm - well, based on these comments, I think this PR isn't really applicable for the best practices, and perhaps we should just document these items in the lsr_role2collection.py script, unless @pcahyna @nhosoi you think there is still something we should have in the general purpose best practices

**Comment by nhosoi** [comment]
> hmm - well, based on these comments, I think this PR isn't really applicable for the best practices, and perhaps we should just document these items in the lsr_role2collection.py script, unless @pcahyna @nhosoi you think there is still something we should have in the general purpose best practices

I agree with you, @richm. Thank you for the inputs, @greg-hellings. I also think our use case is quite special and our findings are not useful for the common collections development. I still think naming modules has to be very careful not to avoid the conflicts, but when implementing something in a collection, it's not an issue at all since the modules are there in the collection and the developer can just use them or add a new one with non-conflict name. :)

File: README.md (hunk 4 above)
Comment (pcahyna) [reply README.md:538]: Sorry for having proposed something that is in conflict with Ansible docs. Still, would it be feasible to change the Ansible standards so that a leading underscore is allowed? We did this for variables in the past. Leading underscores for identifiers are common in programming languages.

File: README.md (hunk 2 above)
Comment (pcahyna) [reply README.md:531]: @greg-hellings @richm I believe it was @t-woerner who argued for such kind of disambiguation using prefixes. IIRC the reason was that as soon as someone uses a `collections:` keyword with our collection, everything gets imported into the namespace and we are back at the old problem with having modules or other plugins without any disambiguation possibly conflicting with each other.
IIRC also similar argument holds for module_utils, but this time the problem occurs even without using the `collections:` keyword. module_utils live in the same namespace, so there is a risk that we will shadow a globally provided module_util with our one. This can be prevented by creating subdirectories under module_utils and putting the actual Python files there. (That said, there is no way to guarantee that the name of subdirectory will not clash with any common module_util, right?)
@t-woerner, can you please confirm that I represented your concerns accurately?

File: README.md (hunk 4 above)
Comment (greg-hellings) [reply README.md:538]: This is not the forum for that sort of change. That needs to be made by the core Ansible Galaxy team, if it is going to be made. It's not a standards restriction, at this point it's literally a technical imitation of the Ansible Galaxy importer.

File: README.md (hunk 2 above)
Comment (greg-hellings) [reply README.md:531]: As with any other place that has namespaces, that collision is a problem for the consumer. Not the producer. If I do `from foo import *` and `from bar import *` and they both have component `baz` then, as the invoker, I call `foo.baz` or `bar.baz` from the point of invocation. I don't go to the authors of foo and bar and demand that they rename to `foo_baz` and `bar_baz`. That's the whole point of namespaces: to allow name collision resolution. It is going against the whole purpose of collection namespacing to further namespace the module name by appending the role that it is used in to the filename. Besides, what is to stop `oasis_roles.system` and `foo.bar` collections from both having role named `do_thing` and thus they both end up with module `do_thing_module`.

You're not back the same problem as before. You now have the FQCN to access the desired module through. Don't further complicate naming schemes with unnecessary bits that are ultimately unproductive and, if you're using the module in more than one place, misleading.

The same does NOT hold for module_utils. They are not shoveled into a flat namespace with all other module_utils added. They would be accessed at `from ansible_collections.oasis_roles.system.plugins.module_utils.some_util_file import my_util_function`. The `oasis_roles.system` is the namespacing provided by my collection. At least this is [how it is documented](https://docs.ansible.com/ansible/devel/dev_guide/migrating_roles.html#update-module-utils-role). So, again, introducing additional namespacing and folder structure under here is possible, if a collection has supremely complex module_utils. But do not force it a priori in an attempt to avoid naming collisions with globally available module_utils or with module_utils that live in another collection.

This whole proposal is attempting to solve a problem which was important back in the pre-collection days but which is the exact problem that collections were created to solve. Don't re-invent the wheel on top of collections again. Just use the collection.

Comment (t-woerner) [reply README.md:531]: > @greg-hellings @richm I believe it was @t-woerner who argued for such kind of disambiguation using prefixes. IIRC the reason was that as soon as someone uses a `collections:` keyword with our collection, everything gets imported into the namespace and we are back at the old problem with having modules or other plugins without any disambiguation possibly conflicting with each other.
> IIRC also similar argument holds for module_utils, but this time the problem occurs even without using the `collections:` keyword. module_utils live in the same namespace, so there is a risk that we will shadow a globally provided module_util with our one. This can be prevented by creating subdirectories under module_utils and putting the actual Python files there. (That said, there is no way to guarantee that the name of subdirectory will not clash with any common module_util, right?)
> @t-woerner, can you please confirm that I represented your concerns accurately?

Yes, that is more or less it. You can still use the full qualified name, but we should make sure that an addition or change within system roles will not easily introduce a conflict with other collections that customers are using. Especially with simple and common names.

Comment (pcahyna) [reply README.md:531]: @t-woerner assuming that this was about the first point (module/plugin names), the second (module_utils placement) seems to be sufficiently resolved. As @greg-hellings says, module/plugin names issue is addressed by the collection structure. You seem to be concerned about using the `collections:` keyword, which is very much like `from ... import *` in Python and brings similar problems. To which the response shall be: don't use the `collections:` keyword then, just as it is better to not use `from ... import *`. Isn't that enough?
